util/dvsim/LsfLauncher.py

```diff
@@ -0,0 +1,379 @@
+
+        assert status in ['P', 'F', 'K']
+        self.status = status
+        if self.exit_code is None:
+            self.exit_code = 0 if status == 'P' else 1
+        if err_msg:
+            self.fail_msg += err_msg
+            log.log(VERBOSE, err_msg)
+        self._post_finish(status)
+        return status
+
+    def poll(self):
+        # It is possible we may have determined the status already.
+        if self.status:
+            return self.status
+
+        if not self.job_script_out_fd:
+            # If job id is not set, the bsub command has not been sent yet.
+            if not self.job_id:
+                return 'D'
+
+            # If the bsub output file is not created, we are still in
+            # dispatched state.
+            if not self.job_script_out.is_file():
+                return "D"
+
+            # We redirect the job's output to the log file, so the job script
+            # output remains empty until the point it finishes. This is a very
+            # quick way to check if the job has completed. If nothing has been
+            # written to the job script output yet, then the job is still
+            # running.
+            if not self.job_script_out.stat().st_size:
+                return "D"
+
+            # If we got to this point,  we can now open the job script output
+            # file for reading.
+            try:
+                self.job_script_out_fd = open(self.job_script_out, "r")
+            except IOError as e:
+                return self._finish(
+                    status="F",
+                    err_msg="ERROR: Failed to open {}\n{}.".format(
+                        self.job_script_out, e))
+
+        # Now that the job has completed, we need to determine it's status.
+        #
+        # If the job successfully launched and it failed, the failure message
+        # will appear in its log file (because of the stderr redirection).
+        # But, in some cases, if there is something wrong with the command
+        # itself, bsub might return immediately with an error message, which
+        # will appear in the job script output file. We want to retrieve that
+        # so that we can report the status accurately.
+        #
+        # At this point, we could run bjobs or bhist to determine the status,
+        # but it has been found to be too slow, expecially when running 1000s
+        # of jobs. Plus, we have to read the job script output anyway to look
+        # for those error messages.
+        #
+        # So we just read this file to determine both, the status and extract
+        # the error message, rather than running bjobs or bhist. But there is
+        # one more complication to deal with - if we read the file now, it is
+        # possible that it may not have been fully updated. We will try reading
+        # it anyway. If we are unable to find what we are looking for, then we
+        # will resume reading it again at the next poll. We will do this upto
```

Comment:
Yep, makes sense. I was thinking that an easy solution was to re-start each time, but it sounds like that causes performance problems. I'm not sure, but maybe bsub does line buffering anyway (so this would never go wrong).